common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -3801,6 +3799,466 @@ static jint NativeCrypto_EVP_AEAD_CTX_open_buf(JNIEnv* env, jclass, jlong evpAea
+        conscrypt::jniutil::throwExceptionFromBoringSSLError(env, "EVP_HPKE_CTX_seal");
+        return {};
+    }
+
+    jbyteArray ciphertextArray = env->NewByteArray(encryptedLen);
+    env->SetByteArrayRegion(ciphertextArray, 0, encryptedLen,
+                            reinterpret_cast<const jbyte*>(encrypted.data()));
+    return ciphertextArray;
+}
+
+const EVP_HPKE_AEAD* getHpkeAead(JNIEnv* env, jint aeadValue) {
+    switch (aeadValue) {
+        case 1:
+            return EVP_hpke_aes_128_gcm();
+        case 2:
+            return EVP_hpke_aes_256_gcm();
+        case 3:
+            return EVP_hpke_chacha20_poly1305();
+        default:
+            conscrypt::jniutil::throwException(env, "java/lang/IllegalArgumentException",
+                                               "AEAD is not supported");
+            return nullptr;
+    }
+}
+
+const EVP_HPKE_KDF* getHpkeKdf(JNIEnv* env, jint kdfValue) {
+    if (kdfValue == 1) {
+        return EVP_hpke_hkdf_sha256();
+    } else {
+        conscrypt::jniutil::throwException(env, "java/lang/IllegalArgumentException",
+                                           "KDF is not supported");
+        return nullptr;
+    }
+}
+
+const EVP_HPKE_KEM* getHpkeKem(JNIEnv* env, jint kemValue) {
+    if (kemValue == 32) {
+        return EVP_hpke_x25519_hkdf_sha256();
+    } else {
+        conscrypt::jniutil::throwException(env, "java/lang/IllegalArgumentException",
+                                           "KEM is not supported");
+        return nullptr;
+    }
+}
+
+static jobject NativeCrypto_EVP_HPKE_CTX_setup_recipient(JNIEnv* env, jclass, jint kemValue,
+                                                         jint kdfValue, jint aeadValue,
+                                                         jbyteArray privateKeyArray,
+                                                         jbyteArray encArray,
+                                                         jbyteArray infoArray) {
+    CHECK_ERROR_QUEUE_ON_RETURN;
+    JNI_TRACE("EVP_HPKE_CTX_setup_recipient(%d, %d, %d, %p, %p, %p)", kemValue, kdfValue, aeadValue,
+              privateKeyArray, encArray, infoArray);
+
+    EVP_HPKE_CTX* ctx = EVP_HPKE_CTX_new();
+    EVP_HPKE_KEY* key = EVP_HPKE_KEY_new();
+
+    const EVP_HPKE_KEM* kem = getHpkeKem(env, kemValue);
+    if (kem == nullptr) {
+        return nullptr;
+    }
+    const EVP_HPKE_KDF* kdf = getHpkeKdf(env, kdfValue);
+    if (kdf == nullptr) {
+        return nullptr;
+    }
+    const EVP_HPKE_AEAD* aead = getHpkeAead(env, aeadValue);
+    if (aead == nullptr) {
+        return nullptr;
+    }
+    if (privateKeyArray == nullptr) {
+        conscrypt::jniutil::throwNullPointerException(env, "privateKeyArray == null");
+        return nullptr;
+    }
+    if (encArray == nullptr) {
+        conscrypt::jniutil::throwNullPointerException(env, "encArray == null");
+        return nullptr;
+    }
+
+    jbyte* privateKey = env->GetByteArrayElements(privateKeyArray, 0);
```

Comment:
Ditto about using ScopedByteArrayRO

Reply:
Done. The code has been updated so `privateKey` uses `ScopedByteArrayRO`